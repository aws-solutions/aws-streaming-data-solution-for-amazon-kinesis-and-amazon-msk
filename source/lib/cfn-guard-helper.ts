/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export class CfnGuardHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: string | string[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push(suppression);
            }
        } else {
            rules.push(suppressions);
        }

        // Ensure that the metadata property exists.
        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        if (resource.cfnOptions.metadata.guard?.SuppressedRules) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.guard.SuppressedRules;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule, rule])
            ).values()
        ];

        resource.cfnOptions.metadata.guard = {
            SuppressedRules: uniqueRules
        }
    }
}
