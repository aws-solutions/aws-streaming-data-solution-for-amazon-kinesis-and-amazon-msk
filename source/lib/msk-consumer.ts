/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as iam from '@aws-cdk/aws-iam';

import { ExecutionRole } from './lambda-role-cloudwatch';
import { CfnNagHelper } from './cfn-nag-helper';

export interface KafkaConsumerProps {
    readonly clusterArn: string;

    readonly batchSize: number;
    readonly startingPosition: lambda.StartingPosition;
    readonly topicName: string;
    readonly enabled: boolean;

    readonly code: lambda.Code;
    readonly timeout: cdk.Duration;
    readonly environmentVariables?: { [key: string]: string };
}

export class KafkaConsumer extends cdk.Construct {
    public readonly Function: lambda.IFunction;
    public readonly EventMapping: lambda.EventSourceMapping;

    private MIN_BATCH_SIZE: number = 1;
    private MAX_BATCH_SIZE: number = 10000;
    private MIN_TIMEOUT_SECONDS: number = 1;

    // According to the Lambda docs (https://docs.aws.amazon.com/lambda/latest/dg/kafka-using-cluster.html#kafka-hosting-how-it-works):
    // For Amazon MSK and self-managed Apache Kafka, the maximum amount of time that Lambda allows a function to run before stopping it is 14 minutes.
    private MAX_TIMEOUT_SECONDS: number = 840;

    constructor(scope: cdk.Construct, id: string, props: KafkaConsumerProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.batchSize)) {
            if (props.batchSize < this.MIN_BATCH_SIZE || props.batchSize > this.MAX_BATCH_SIZE) {
                throw new Error(`batchSize must be between ${this.MIN_BATCH_SIZE} and ${this.MAX_BATCH_SIZE} (given ${props.batchSize})`);
            }
        }

        const timeoutSeconds = props.timeout.toSeconds();
        if (timeoutSeconds < this.MIN_TIMEOUT_SECONDS || timeoutSeconds > this.MAX_TIMEOUT_SECONDS) {
            throw new Error(`timeout must be a value between ${this.MIN_TIMEOUT_SECONDS} and ${this.MAX_TIMEOUT_SECONDS} seconds (given ${timeoutSeconds})`);
        }

        const executionRole = new ExecutionRole(this, 'Role', {
            inlinePolicyName: 'MskPolicy',
            inlinePolicyDocument: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({
                        actions: [
                            'kafka:DescribeCluster',
                            'kafka:GetBootstrapBrokers',
                            'ec2:CreateNetworkInterface',
                            'ec2:DescribeNetworkInterfaces',
                            'ec2:DescribeVpcs',
                            'ec2:DeleteNetworkInterface',
                            'ec2:DescribeSubnets',
                            'ec2:DescribeSecurityGroups'
                        ],
                        resources: ['*']
                    })
                ]
            })
        });

        const cfnRole = executionRole.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'Actions do not support resource level permissions'
        });

        this.Function = new lambda.Function(this, 'Consumer', {
            runtime: lambda.Runtime.NODEJS_14_X,
            handler: 'index.handler',
            role: executionRole.Role,
            code: props.code,
            timeout: props.timeout,
            environment: props.environmentVariables
        });

        this.EventMapping = this.Function.addEventSourceMapping('Mapping', {
            eventSourceArn: props.clusterArn,
            startingPosition: props.startingPosition,
            batchSize: props.batchSize,
            kafkaTopic: props.topicName,
            enabled: props.enabled
        });
    }
}
