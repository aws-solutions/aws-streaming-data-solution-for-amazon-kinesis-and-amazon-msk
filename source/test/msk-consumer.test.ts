/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { aws_lambda as lambda } from 'aws-cdk-lib';
import { Template} from 'aws-cdk-lib/assertions';

import { KafkaConsumer } from '../lib/msk-consumer';

describe('successful scenarios', () => {
    let stack: cdk.Stack;
    let lambdaCodeParams: lambda.CfnParametersCodeProps;

    beforeEach(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack1');

        lambdaCodeParams = {
            bucketNameParam: new cdk.CfnParameter(stack, 'LambdaCodeBucket'),
            objectKeyParam: new cdk.CfnParameter(stack, 'LambdaCodeKey'),
        };
    });

    test('creates a MSK Lambda consumer', () => {
        new KafkaConsumer(stack, 'TestMsk', {
            clusterArn: 'arn:aws:kafka:region:account:cluster/cluster-name/cluster-uuid',
            scramSecretArn: 'my-secret-arn',
            batchSize: 10,
            enabled: true,
            startingPosition: lambda.StartingPosition.LATEST,
            timeout: cdk.Duration.minutes(1),
            code: lambda.Code.fromCfnParameters(lambdaCodeParams),
            topicName: 'my-topic',
            environmentVariables: {
                'MY_ENV_VARIABLE': 'foo'
            }
        });

        Template.fromStack(stack).hasResource('AWS::IAM::Role', {
            Metadata: {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W11',
                        reason: 'Actions do not support resource level permissions'
                    }]
                }
            }
        });
    });

    test('accepts CloudFormation parameters', () => {
        const clusterArn = new cdk.CfnParameter(stack, 'ClusterArn', { type: 'String' });
        const secretArn = new cdk.CfnParameter(stack, 'SecretArn', { type: 'String' });
        const batchSize = new cdk.CfnParameter(stack, 'BatchSize', { type: 'Number' });
        const topicName = new cdk.CfnParameter(stack, 'TopicName', { type: 'String' });

        new KafkaConsumer(stack, 'TestMsk', {
            clusterArn: clusterArn.valueAsString,
            scramSecretArn: secretArn.valueAsString,
            batchSize: batchSize.valueAsNumber,
            enabled: true,
            startingPosition: lambda.StartingPosition.LATEST,
            timeout: cdk.Duration.minutes(1),
            code: lambda.Code.fromCfnParameters(lambdaCodeParams),
            topicName: topicName.valueAsString
        });

        Template.fromStack(stack).hasResourceProperties('AWS::Lambda::EventSourceMapping', {
            EventSourceArn: { Ref: 'ClusterArn' },
            BatchSize: { Ref: 'BatchSize' },
            Topics: [{ Ref: 'TopicName' }],
            SourceAccessConfigurations: [{
                Type: 'SASL_SCRAM_512_AUTH',
                URI: { Ref: 'SecretArn' }
            }]
        });
    });

describe('validation tests', () => {
    let stack: cdk.Stack;
    let lambdaCodeParams: lambda.CfnParametersCodeProps;

    beforeEach(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack2');

        lambdaCodeParams = {
            bucketNameParam: new cdk.CfnParameter(stack, 'LambdaCodeBucket'),
            objectKeyParam: new cdk.CfnParameter(stack, 'LambdaCodeKey'),
        };
    });

    test.each([0, 901])('timeout must be between allowed values', (invalidTimeoutSeconds) => {
        expect(() => new KafkaConsumer(stack, 'TestMsk', {
            clusterArn: 'arn:aws:kafka:region:account:cluster/cluster-name/cluster-uuid',
            batchSize: 10,
            enabled: true,
            startingPosition: lambda.StartingPosition.LATEST,
            timeout: cdk.Duration.seconds(invalidTimeoutSeconds),
            code: lambda.Code.fromCfnParameters(lambdaCodeParams),
            topicName: 'my-topic'
        })).toThrowError(`timeout must be a value between 1 and 900 seconds (given ${invalidTimeoutSeconds})`);
    });

    test.each([-1, 0, 10001])('batch size must be between allowed values', (invalidBatchSize) => {
        expect(() => new KafkaConsumer(stack, 'TestMsk', {
            clusterArn: 'arn:aws:kafka:region:account:cluster/cluster-name/cluster-uuid',
            batchSize: invalidBatchSize,
            enabled: true,
            startingPosition: lambda.StartingPosition.LATEST,
            timeout: cdk.Duration.minutes(1),
            code: lambda.Code.fromCfnParameters(lambdaCodeParams),
            topicName: 'my-topic'
        })).toThrowError(`batchSize must be between 1 and 10000 (given ${invalidBatchSize})`);
    });
})
});
