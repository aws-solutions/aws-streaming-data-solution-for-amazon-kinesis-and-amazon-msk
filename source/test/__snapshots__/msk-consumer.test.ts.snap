// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful scenarios accepts CloudFormation parameters 1`] = `
Object {
  "Conditions": Object {
    "TestMskIsSecretEmptyConditionD1E36ADB": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "SecretArn",
        },
        "",
      ],
    },
    "TestMskIsSecretNotEmptyCondition9929A158": Object {
      "Fn::Not": Array [
        Object {
          "Condition": "TestMskIsSecretEmptyConditionD1E36ADB",
        },
      ],
    },
  },
  "Parameters": Object {
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292ArtifactHash8E89EA8E": Object {
      "Description": "Artifact hash for asset \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3BucketD5D1EAB7": Object {
      "Description": "S3 bucket for asset \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C": Object {
      "Description": "S3 key for asset version \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "BatchSize": Object {
      "Type": "Number",
    },
    "ClusterArn": Object {
      "Type": "String",
    },
    "LambdaCodeBucket": Object {
      "Type": "String",
    },
    "LambdaCodeKey": Object {
      "Type": "String",
    },
    "SecretArn": Object {
      "Type": "String",
    },
    "TopicName": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestMskConsumer4592A744": Object {
      "DependsOn": Array [
        "TestMskRole86578511",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "LambdaCodeBucket",
          },
          "S3Key": Object {
            "Ref": "LambdaCodeKey",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestMskRole86578511",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestMskConsumerMapping79C4A58A": Object {
      "Condition": "TestMskIsSecretEmptyConditionD1E36ADB",
      "Properties": Object {
        "BatchSize": Object {
          "Ref": "BatchSize",
        },
        "Enabled": true,
        "EventSourceArn": Object {
          "Ref": "ClusterArn",
        },
        "FunctionName": Object {
          "Ref": "TestMskConsumer4592A744",
        },
        "StartingPosition": "LATEST",
        "Topics": Array [
          Object {
            "Ref": "TopicName",
          },
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TestMskConsumerMappingWithSecretBF6972F1": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DependsOn": Array [
        "TestMskSecretPolicy272AB2E3",
      ],
      "Properties": Object {
        "BatchSize": Object {
          "Ref": "BatchSize",
        },
        "Enabled": true,
        "EventSourceArn": Object {
          "Ref": "ClusterArn",
        },
        "FunctionName": Object {
          "Ref": "TestMskConsumer4592A744",
        },
        "SourceAccessConfigurations": Array [
          Object {
            "Type": "SASL_SCRAM_512_AUTH",
            "URI": Object {
              "Ref": "SecretArn",
            },
          },
        ],
        "StartingPosition": "LATEST",
        "Topics": Array [
          Object {
            "Ref": "TopicName",
          },
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TestMskCustomResourceCD43A296": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DependsOn": Array [
        "TestMskCustomResourceRoleDA0E0EAF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3BucketD5D1EAB7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestMskCustomResourceRoleDA0E0EAF",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestMskCustomResourceRoleDA0E0EAF": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "secretsmanager:DescribeSecret",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Ref": "SecretArn",
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "SecretMetadataPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestMskRole86578511": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "Actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kafka:DescribeCluster",
                    "kafka:GetBootstrapBrokers",
                    "kafka:ListScramSecrets",
                    "ec2:CreateNetworkInterface",
                    "ec2:DescribeNetworkInterfaces",
                    "ec2:DescribeVpcs",
                    "ec2:DeleteNetworkInterface",
                    "ec2:DescribeSubnets",
                    "ec2:DescribeSecurityGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "MskPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestMskSecretMetadataF8DCB6CD": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "SecretArn": Object {
          "Ref": "SecretArn",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestMskCustomResourceCD43A296",
            "Arn",
          ],
        },
      },
      "Type": "Custom::SecretMetadata",
      "UpdateReplacePolicy": "Delete",
    },
    "TestMskSecretPolicy272AB2E3": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "kms:Decrypt",
              "Condition": Object {
                "StringEquals": Object {
                  "kms:ViaService": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "secretsmanager.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestMskSecretMetadataF8DCB6CD",
                  "KmsKeyId",
                ],
              },
            },
            Object {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "SecretArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestMskSecretPolicy272AB2E3",
        "Roles": Array [
          Object {
            "Ref": "TestMskRole86578511",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`successful scenarios creates a MSK Lambda consumer 1`] = `
Object {
  "Conditions": Object {
    "TestMskIsSecretEmptyConditionD1E36ADB": Object {
      "Fn::Equals": Array [
        "my-secret-arn",
        "",
      ],
    },
    "TestMskIsSecretNotEmptyCondition9929A158": Object {
      "Fn::Not": Array [
        Object {
          "Condition": "TestMskIsSecretEmptyConditionD1E36ADB",
        },
      ],
    },
  },
  "Parameters": Object {
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292ArtifactHash8E89EA8E": Object {
      "Description": "Artifact hash for asset \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3BucketD5D1EAB7": Object {
      "Description": "S3 bucket for asset \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C": Object {
      "Description": "S3 key for asset version \\"46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292\\"",
      "Type": "String",
    },
    "LambdaCodeBucket": Object {
      "Type": "String",
    },
    "LambdaCodeKey": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestMskConsumer4592A744": Object {
      "DependsOn": Array [
        "TestMskRole86578511",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "LambdaCodeBucket",
          },
          "S3Key": Object {
            "Ref": "LambdaCodeKey",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "MY_ENV_VARIABLE": "foo",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestMskRole86578511",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestMskConsumerMapping79C4A58A": Object {
      "Condition": "TestMskIsSecretEmptyConditionD1E36ADB",
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": true,
        "EventSourceArn": "my-cluster-arn",
        "FunctionName": Object {
          "Ref": "TestMskConsumer4592A744",
        },
        "StartingPosition": "LATEST",
        "Topics": Array [
          "my-topic",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TestMskConsumerMappingWithSecretBF6972F1": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DependsOn": Array [
        "TestMskSecretPolicy272AB2E3",
      ],
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": true,
        "EventSourceArn": "my-cluster-arn",
        "FunctionName": Object {
          "Ref": "TestMskConsumer4592A744",
        },
        "SourceAccessConfigurations": Array [
          Object {
            "Type": "SASL_SCRAM_512_AUTH",
            "URI": "my-secret-arn",
          },
        ],
        "StartingPosition": "LATEST",
        "Topics": Array [
          "my-topic",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TestMskCustomResourceCD43A296": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DependsOn": Array [
        "TestMskCustomResourceRoleDA0E0EAF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3BucketD5D1EAB7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters46082e48b826df3817b5cfafd5bfc42d905840ad65d894855ec58966925e1292S3VersionKey12D5174C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestMskCustomResourceRoleDA0E0EAF",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestMskCustomResourceRoleDA0E0EAF": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "secretsmanager:DescribeSecret",
                  "Effect": "Allow",
                  "Resource": "my-secret-arn",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "SecretMetadataPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestMskRole86578511": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "Actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kafka:DescribeCluster",
                    "kafka:GetBootstrapBrokers",
                    "kafka:ListScramSecrets",
                    "ec2:CreateNetworkInterface",
                    "ec2:DescribeNetworkInterfaces",
                    "ec2:DescribeVpcs",
                    "ec2:DeleteNetworkInterface",
                    "ec2:DescribeSubnets",
                    "ec2:DescribeSecurityGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "MskPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestMskSecretMetadataF8DCB6CD": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "SecretArn": "my-secret-arn",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestMskCustomResourceCD43A296",
            "Arn",
          ],
        },
      },
      "Type": "Custom::SecretMetadata",
      "UpdateReplacePolicy": "Delete",
    },
    "TestMskSecretPolicy272AB2E3": Object {
      "Condition": "TestMskIsSecretNotEmptyCondition9929A158",
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "kms:Decrypt",
              "Condition": Object {
                "StringEquals": Object {
                  "kms:ViaService": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "secretsmanager.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestMskSecretMetadataF8DCB6CD",
                  "KmsKeyId",
                ],
              },
            },
            Object {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": "my-secret-arn",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestMskSecretPolicy272AB2E3",
        "Roles": Array [
          Object {
            "Ref": "TestMskRole86578511",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
